### PR TITLE
Add new DataType class and use to support typed literals (for 468)

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
I have one un-covered line in the diff but it's in dynamo0p1.py and it wasn't covered before either so I propose to leave it. Ready for first review, probably from @sergisiso or @rupertford but @hiker could also do it :-)

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -6117,26 +6117,88 @@ def reference_accesses(self, var_accesses):
...
     @property
-    def value(self):
+    def datatype(self):
+        '''
+        :returns: the type of this Literal.
+        :rtype: :py:class:`psyclone.psyGen.DataType`
+        '''
+        return self._datatype
+
+    @datatype.setter
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: Is this setter used anywhere, apart from the __init__ of this class? Similary for the value setter. If not, perhaps it makes sense for Literals to be immutable?

**Comment by rupertford** [comment]
Back to @arporter

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -6117,26 +6117,88 @@ def reference_accesses(self, var_accesses):
 
 class Literal(Node):
     '''
-    Node representing a Literal
+    Node representing a Literal.
 
-    :param str value: String representing the literal value.
+    :param value: the value of the literal.
+    :type value: str (for numberical values) or bool
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Oops.

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -6117,26 +6117,88 @@ def reference_accesses(self, var_accesses):
...
+    :param datatype: The datatype of this literal.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Oops.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -6117,26 +6117,88 @@ def reference_accesses(self, var_accesses):
...
+    def datatype(self, value):
         '''
-        Return the value of the literal.
+        Setter for the data-type of this Literal.
 
+        :param value: the data-type.
+        :type value: :py:class:`psyclone.psyir.nodes.DataType`
+
+        :raises TypeError: if the value is not an instance of \
+                           :py:class:`psyclone.psyir.nodes.DataType`
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Yes.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 5]
```diff
@@ -46,14 +46,21 @@
...
+from psyclone.psyir.nodes import DataType, TYPE_MAP_TO_PYTHON
+from psyclone.psyir.symbols import SymbolError, DataSymbol, ContainerSymbol, \
     GlobalInterface, ArgumentInterface, UnresolvedInterface, LocalInterface
 
 # The list of Fortran instrinsic functions that we know about (and can
 # therefore distinguish from array accesses). These are taken from
 # fparser.
 FORTRAN_INTRINSICS = Fortran2003.Intrinsic_Name.function_names
 
+# Mapping from Fortran data types to PSyIR types
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: Done.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 6]
```diff
@@ -1979,11 +1989,52 @@ def _number_handler(self, node, parent):
...
+            # Unrecognised datatype - will result in a CodeBlock
+            raise NotImplementedError()
+
+    def _char_literal_handler(self, node, parent):
+        '''
+        Transforms an fparser2 character literal into a PSyIR literal.
+
+        :param node: node in fparser2 parse tree.
+        :type node: :py:class:`fparser.two.Fortran2003.Char_Literal_Constant`
+        :param parent: Parent node of the PSyIR node we are constructing.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: Done.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 7]
```diff
@@ -1979,11 +1989,52 @@ def _number_handler(self, node, parent):
...
+        '''
+        return Literal(str(node.items[0]), DataType.CHARACTER, parent=parent)
+
+    def _bool_literal_handler(self, node, parent):
+        '''
+        Transforms an fparser2 logical literal into a PSyIR literal.
+
+        :param node: node in fparser2 parse tree.
+        :type node: :py:class:`fparser.two.Fortran2003.Char_Literal_Constant`
+        :param parent: Parent node of the PSyIR node we are constructing.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: Yep.

File: src/psyclone/psyir/symbols/datasymbol.py [hunk 8]
```diff
@@ -73,25 +74,10 @@ class DataSymbol(Symbol):
     :type precision: int or :py:class:`psyclone.psyir.symbol.DataSymbol` or \
                      :py:class:`psyclone.psyir.symbols.DataSymbol.Precision`
 
-    :raises NotImplementedError: provided parameters are not supported yet.
-    :raises TypeError: provided parameters have invalid error type.
+    :raises TypeError: provided parameters have invalid type.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/datasymbol.py]: Done.

File: src/psyclone/dynamo0p1.py [hunk 9]
```diff
@@ -257,11 +260,11 @@ def load(self, kern):
     def gen_code(self, parent):
         ''' Work out the appropriate loop bounds and then call the base
             class to generate the code '''
-        self.start_expr = Literal("1", parent=self)
+        self.start_expr = Literal("1", DataType.INTEGER, parent=self)
         if self._loop_type == "colours":
             self.stop_expr = Reference("ncolour", parent=self)
         elif self._loop_type == "colour":
-            self.stop_expr = ArrayReference("ncp_ncolour", parent=self)
+            self.stop_expr = Array("ncp_ncolour", parent=self)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p1.py]: Yes, ArrayReference doesn't exist.

File: src/psyclone/psyir/nodes/datatypes.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the datatype definitions.'''
+
+from enum import Enum
+
+
+class DataType(Enum):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/datatypes.py]: OK.

File: src/psyclone/psyGen.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/psyGen.py]: Yes, good idea.

**Comment by arporter** [comment]
All done. Moving DataType into symbols feels good and means I won't clash with @hiker over the `psyir/nodes` directory. Ready for next review if Travis is happy.

File: src/psyclone/tests/psyir/backend/sir_test.py [hunk 11]
```diff
@@ -657,7 +657,8 @@ def test_sirwriter_literal_node_1(parser, sir_writer):
 
 
 # (2/2) Method literal_node
-@pytest.mark.xfail(reason="#468 PSyIR does not capture the type of literals")
+@pytest.mark.xfail(reason="#612 SIR backend does not yet make use of the"
```
Comment (rupertford) [inline src/psyclone/tests/psyir/backend/sir_test.py]: Please add modified by A. R. Porter.

File: src/psyclone/psyir/backend/fortran.py [hunk 12]
```diff
@@ -51,6 +52,13 @@
 # fparser.
 FORTRAN_INTRINSICS = Fortran2003.Intrinsic_Name.function_names
 
+# Mapping from PSyIR types to Fortran data types - simply reverse the map
+# from the frontend.
+TYPE_MAP_TO_FORTRAN = {}
+for key, item in TYPE_MAP_FROM_FORTRAN.items():
+    TYPE_MAP_TO_FORTRAN[item] = key
+
+
```
Comment (rupertford) [inline src/psyclone/psyir/backend/fortran.py]: pycodestyle error - too many lines.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 13]
```diff
@@ -136,12 +136,12 @@ def test_gen_datatype(datatype, result):
...
-     ("integer", 8, "integer*8"),
-     ("real", 16, "real*16"),
-     ("real", DataSymbol.Precision.DOUBLE, "double precision"),
-     ("integer", DataSymbol("i_def", "integer"), "integer(kind=i_def)"),
-     ("real", DataSymbol("r_def", "integer"), "real(kind=r_def)")])
+    [(DataType.REAL, None, "real"),
+     (DataType.INTEGER, 8, "integer*8"),
+     (DataType.REAL, 16, "real*16"),
+     (DataType.REAL, DataSymbol.Precision.DOUBLE, "double precision"),
+     (DataType.INTEGER, DataSymbol("i_def", DataType.INTEGER), "integer(kind=i_def)"),
```
Comment (rupertford) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: pycodestyle: Lines too long

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 14]
```diff
@@ -163,7 +163,7 @@ def test_gen_datatype_precision(datatype, precision, result):
 #     '''
 #     import logging
 #     with caplog.at_level(logging.WARNING):
-#         symbol = Symbol("dummy", "integer",precision=Symbol.Precision.DOUBLE)
+#         symbol = Symbol("dummy", DataType.INTEGER,precision=Symbol.Precision.DOUBLE)
```
Comment (rupertford) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: pycodestyle: Line too long

**Comment by rupertford** [comment]
Back to @arporter - very close now.

File: src/psyclone/tests/psyir/frontend/literals_test.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+from fparser.common.readfortran import FortranStringReader
+from fparser.two import Fortran2003
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+from psyclone.psyir.symbols import DataType
+from psyclone.psyGen import Node, Literal, CodeBlock
+
+
+@pytest.mark.parametrize("code, dtype", [("'hello'", DataType.CHARACTER),
+                                         ("1", DataType.INTEGER),
+                                         ("1", DataType.INTEGER),
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/frontend/literals_test.py]: Oh yes, removed.

File: src/psyclone/psyir/backend/c.py [hunk 16]
```diff
@@ -42,6 +42,14 @@
 '''
 
 from psyclone.psyir.backend.visitor import PSyIRVisitor, VisitorError
+from psyclone.psyir.symbols import DataType
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/c.py]: Done.

File: src/psyclone/psyir/backend/sir.py [hunk 17]
```diff
@@ -403,7 +403,7 @@ def literal_node(self, node):
         '''
         result = node.value
         # There is an assumption here that the literal is a float (see
-        # #468).
+        # #612).
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/sir.py]: Done

File: src/psyclone/tests/psyir/backend/sir_test.py (hunk 11 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/sir_test.py]: Done

**Comment by rupertford** [comment]
Oh, and I forgot to say. Could you please update fixtures in working practices in the developers guide with the f2008 fixture.

File: src/psyclone/psyir/backend/fortran.py (hunk 12 above)
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Oops.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 13 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Apologies - done.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 14 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Oops.

**Comment by arporter** [comment]
Should be all good now. Apologies for missing those pylint/pep8 errors. I've updated the developer guide as requested. Ready for next review once Travis is happy.

**Pull request merged**